crawler/file_processing.py

```diff
@@ -663,60 +718,395 @@ def format_and_filter_rows(self, csvreader: DictReader) -> Any:
+
+        if not self.is_row_channel_cq_valid(row, line_number, FIELD_CH4_CQ):
+            return False
+
+        return True
+
+    def is_within_cq_range(self, range_min, range_max, num) -> bool:
+        """Validation to check if a number lies within the expected range.
+
+        Arguments:
+            range_min {Decimal} - minimum range number
+            range_max {Decimal} - maximum range number
+            num {str} - string representation of the number to be tested
+
+        Returns:
+            bool - whether the value lies within range
+        """
+        return range_min <= Decimal(num) <= range_max
+
+    def is_row_channel_cq_in_range(self, row, line_number, fieldname) -> bool:
+        """Is the channel cq within the specified range.
+
+        Arguments:
+            row {Row} - row object from the csvreader
+            line_number {integer} - line number within the file
+            fieldname {str} - the name of the cq column
+
+        Returns:
+            bool - whether the cq value is valid
+        """
+        if fieldname in row and row[fieldname]:
+            if not self.is_within_cq_range(MIN_CQ_VALUE, MAX_CQ_VALUE, row[fieldname]):
+                self.logging_collection.add_error(
+                    "TYPE 20",
+                    f"{fieldname} not in range ({MIN_CQ_VALUE}, {MAX_CQ_VALUE}), line: {line_number}, result: {row[fieldname]}",
+                )
+                return False
+
+        return True
+
+    def row_channel_cq_values_within_range(self, row, line_number) -> bool:
+        """Validation to check if the row channel cq values are within range.
+
+        Arguments:
+            row {Row} - row object from the csvreader
+            line_number {integer} - line number within the file
+
+        Returns:
+            bool - whether the cq values are within range
+        """
+        if not self.is_row_channel_cq_in_range(row, line_number, FIELD_CH1_CQ):
+            return False
+
+        if not self.is_row_channel_cq_in_range(row, line_number, FIELD_CH2_CQ):
+            return False
+
+        if not self.is_row_channel_cq_in_range(row, line_number, FIELD_CH3_CQ):
+            return False
+
+        if not self.is_row_channel_cq_in_range(row, line_number, FIELD_CH4_CQ):
+            return False
+
+        return True
+
+    def row_positive_result_matches_channel_results(self, row, line_number) -> bool:
+        """Validation to check that when the result is positive, and channel results are present,
+           then at least one of the channel results is also positive.
+
+        Arguments:
+            row {Row} - row object from the csvreader
+            line_number {integer} - line number within the file
+
+        Returns:
+            bool - whether the channel results complement the main results
+        """
+        # if the result is not positive we do not need to check any further
+        if row[FIELD_RESULT] != 'Positive':
+            return True
```

Comment:
OK still got a few comments about this validation, sorry! (just from reading JB's comments on the story)

1.
Think another disallowed combo would be if: \
Result = 'Negative'
Any of the CHn-Result values = 'Positive'
?

2. Don't think you should be using the FIELD_CH4_RESULT at all, just 1-3